pkg/jobmanager/job.go

```diff
@@ -74,28 +89,53 @@ func (j *Job) ExitCode() int {
 	return c
 }
 
-// wait for the underlying process. Broadcast the end via waitChan
-// and close the given resources.
-func (j *Job) wait() {
-	if err := j.cmd.Wait(); err != nil {
-		slog.Debug("Process Wait ended with error", "error", err)
-	}
-
+func (j *Job) Close() {
 	j.mu.Lock()
-	defer j.mu.Unlock()
 	if j.status != pb.JobStatus_JOB_STATUS_STOPPED {
 		j.status = pb.JobStatus_JOB_STATUS_EXITED
 	}
-	j.exitCode = j.cmd.ProcessState.ExitCode()
+	if j.cmd.ProcessState != nil {
+		j.exitCode = j.cmd.ProcessState.ExitCode()
+	}
 	close(j.waitChan)
-	if err := j.broadcaster.Close(); err != nil {
+	if e1 := j.broadcaster.Close(); e1 != nil {
 		// Best effort.
-		slog.Error("Broadcaster closed with error.", "error", err)
+		slog.Error("Broadcaster closed with error.", "error", e1)
 	}
+	j.mu.Unlock()
+
+	// NOTE: cgroupPath is immutable and set at start before being shared, can
+	// safely be used without lock.
+	if err := os.RemoveAll(j.cgroupPath); err != nil {
+		// Best effort.
```

Comment:
Is waitpid on the init of a pid namespace guaranteed to return after all the processes in the namespace are fully cleaned up? Even ignoring the potential for other things to be put in the cgroup by something else (which violates the cgroup single writer principle) I think that there can be some time between the termination of the main process and the cgroup becoming unpopulated.